### PR TITLE
join

### PR DESCRIPTION
**Pull request by gdementen**
WIP implementation of array.align and all that was necessary
(AxisCollection.align, Axis.align)

### CODE REVIEW

**Comment by gdementen** [comment]
also the choice of default join method is not innocent and should be considered carefully, as this should be consistent with what we will do for binary ops (which should be almost trivial to implement once this PR is merged).

File: larray/core.py [hunk 1]
```diff
@@ -1830,6 +1830,45 @@ def difference(self, other):
...
+        >>> axis1.align(axis2)
+        Axis(['a0', 'a1', 'a2', 'a3'], 'a')
+        >>> axis1.align(axis2, join='inner')
+        Axis(['a1', 'a2'], 'a')
+        >>> axis1.align(axis2, join='left')
+        Axis(['a0', 'a1', 'a2'], 'a')
+        >>> axis1.align(axis2, join='right')
+        Axis(['a1', 'a2', 'a3'], 'a')
+        """
+        assert join in {'outer', 'inner', 'left', 'right'}
```
Comment (gdementen) [inline larray/core.py]: I should raise a ValueError exception instead of using an assert

File: larray/core.py [hunk 2]
```diff
@@ -3762,6 +3801,63 @@ def split_axis(self, axis, sep='_', names=None, regex=None):
...
+        ])
+        >>> aligned2
+        AxisCollection([
+            Axis(['a0', 'a1', 'a2'], 'a'),
+            Axis(['b0', 'b1', 'b2'], 'b')
+        ])
+        """
+        assert join in {'outer', 'inner', 'left', 'right'}
+        other = other if isinstance(other, AxisCollection) else AxisCollection(other)
+        join_axes = list(OrderedSet(self.names) & OrderedSet(other.names)) if axes is None else axes
```
Comment (alixdamman) [inline larray/core.py]: What about anonymous axes? Did you try what happen in case of anonymous axes?

File: larray/core.py [hunk 3]
```diff
@@ -5069,6 +5165,76 @@ def get_labels(self_axis):
...
+        a1 |  b0 | 5.0
+        a1 |  b1 | 7.0
+        a1 |  b2 | nan
+        a2 |  b0 | nan
+        a2 |  b1 | nan
+        a2 |  b2 | nan
+        """
+        other = aslarray(other)
+        left_axes, right_axes = self.axes.align(other.axes, join=join, axes=axes)
+        return self.reindex(left_axes, fill_value=fill_value), other.reindex(right_axes, fill_value=fill_value)
```
Comment (alixdamman) [inline larray/core.py]: If I understand well, AxisCollection.align returns the same AxisCollection `new_axes` twice. So why not returning only one `new_axes` collection and replace here left_axes, right_axes by common_axes?

File: larray/core.py (hunk 2 above)
Comment (gdementen) [inline larray/core.py]: Not yet. I will certainly break at this point. The way to handle them is relatively easy when there are only anonymous axes (ie do it by axis position) but when mixing anonymous and non-anonymous, this gets tricky and I don't really know what we should do.

File: larray/core.py (hunk 3 above)
Comment (gdementen) [inline larray/core.py]: No. Please see the comment at https://github.com/liam2/larray/pull/254/files#diff-d2e8c0a8b16062fcfdf68b1de7e842caR3857

**Comment by alixdamman** [comment]
should be referenced in changelog 0.23

**Comment by gdementen** [comment]
* I am still unsure about the default value for join here
* it still does not work for anonymous axes, but I am unsure we should care at this point

File: doc/source/changes/version_0_25.rst.inc [hunk 4]
```diff
@@ -5,6 +5,67 @@
   It contains a link to open the documentation, a shortcut to launch the user interface in edition mode
   and a shortcut to update larrayenv. Closes issue:`281`.
 
+* implemented array.align(other_array) which makes two arrays compatible with each other (by making all common axes
+  compatible). This is done by adding, removing or reordering labels for each common axis according to the join method:
+
+    - outer: will use a label if it is in either arrays axis (ordered like the first array). This is the default.
```
Comment (alixdamman) [inline doc/source/changes/version_0_25.rst.inc]: This is the default option as it results in zero information loss.

File: doc/source/changes/version_0_25.rst.inc [hunk 5]
```diff
@@ -5,6 +5,67 @@
   It contains a link to open the documentation, a shortcut to launch the user interface in edition mode
   and a shortcut to update larrayenv. Closes issue:`281`.
 
+* implemented array.align(other_array) which makes two arrays compatible with each other (by making all common axes
+  compatible). This is done by adding, removing or reordering labels for each common axis according to the join method:
```
Comment (alixdamman) [inline doc/source/changes/version_0_25.rst.inc]: ... compatible). The fill value for missing labels defaults to nan. This is done by ...

Comment (alixdamman) [inline doc/source/changes/version_0_25.rst.inc]: according to the join methodS?

File: doc/source/changes/version_0_25.rst.inc [hunk 6]
```diff
@@ -5,6 +5,67 @@
...
+    - left: will use the first array axis labels
+    - right: will use the other array axis labels
+
+    >>> arr1 = ndtest((2, 3))
+    >>> arr1
+    a\b  b0  b1  b2
+     a0   0   1   2
+     a1   3   4   5
+    >>> arr2 = -ndtest((3, 2))
+    >>> # reorder array to make the test more interesing
```
Comment (alixdamman) [inline doc/source/changes/version_0_25.rst.inc]: interesTing

File: doc/source/changes/version_0_25.rst.inc [hunk 7]
```diff
@@ -5,6 +5,67 @@
...
+    >>> arr2 = arr2[['b1', 'b0']]
+    >>> arr2
+    a\\b  b1  b0
+     a0  -1   0
+     a1  -3  -2
+     a2  -5  -4
```
Comment (alixdamman) [inline doc/source/changes/version_0_25.rst.inc]: /blankline
align arrays arr1 and arr2
/blankline

File: larray/core/array.py [hunk 8]
```diff
@@ -1473,14 +1473,175 @@ def get_labels(self_axis):
         else:
             return res
 
+    def align(self, other, join='outer', fill_value=nan, axes=None):
```
Comment (alixdamman) [inline larray/core/array.py]: I would keep 'outer' as default option but maybe add a warning in the doctrsing.

File: larray/core/array.py [hunk 9]
```diff
@@ -1473,14 +1473,175 @@ def get_labels(self_axis):
...
+
+        Parameters
+        ----------
+        other : LArray-like
+        join : {'outer', 'inner', 'left', 'right'}, optional
+            Join method. For each axis common to both arrays:
+              - outer (default): will use a label if it is in either arrays axis (ordered like the first array)
+              - inner: will use a label if it is in both arrays axis (ordered like the first array)
+              - left: will use the first array axis labels
+              - right: will use the other array axis labels.
```
Comment (alixdamman) [inline larray/core/array.py]: Defaults to 'outer' as it results in zero information loss.

File: larray/core/array.py [hunk 10]
```diff
@@ -1473,14 +1473,175 @@ def get_labels(self_axis):
...
+        fill_value : scalar or LArray, optional
+            Value to use for missing values. Defaults to NaN.
```
Comment (alixdamman) [inline larray/core/array.py]: nan?

File: larray/core/array.py [hunk 11]
```diff
@@ -1473,14 +1473,175 @@ def get_labels(self_axis):
...
+
+        Examples
+        --------
+        >>> arr1 = ndtest((2, 3))
+        >>> arr1
+        a\\b  b0  b1  b2
+         a0   0   1   2
+         a1   3   4   5
+        >>> arr2 = -ndtest((3, 2))
+        >>> # reorder array to make the test more interesing
```
Comment (alixdamman) [inline larray/core/array.py]: same as in changelog

File: larray/core/array.py [hunk 12]
```diff
@@ -1473,14 +1473,175 @@ def get_labels(self_axis):
...
+        >>> arr2 = arr2[['b1', 'b0']]
+        >>> arr2
+        a\\b  b1  b0
+         a0  -1   0
+         a1  -3  -2
+         a2  -5  -4
```
Comment (alixdamman) [inline larray/core/array.py]: same as in changelog

File: larray/core/array.py (hunk 8 above)
Comment (alixdamman) [inline larray/core/array.py]: Need also a comment about anonymous axes (maybe a doctest that raises an Error).

File: larray/core/axis.py [hunk 13]
```diff
@@ -998,6 +998,45 @@ def difference(self, other):
...
+        ----------
+        other : Axis or label sequence
+        join : {'outer', 'inner', 'left', 'right'}, optional
+            Defaults to 'outer'.
+
+        Returns
+        -------
+        Axis
+            Aligned axis
+
```
Comment (alixdamman) [inline larray/core/axis.py]: See Also
--------
LArray.align

File: larray/core/axis.py [hunk 14]
```diff
@@ -2200,6 +2242,65 @@ def split_axis(self, axis, sep='_', names=None, regex=None):
...
+            Axes to align. Need to be valid in both arrays. Defaults to None (all common axes).
+
+        Returns
+        -------
+        (left, right) : (AxisCollection, AxisCollection)
+            Aligned collections
+
+        Examples
+        --------
+
```
Comment (alixdamman) [inline larray/core/axis.py]: No need for this blankline after Examples

File: larray/core/axis.py [hunk 15]
```diff
@@ -2200,6 +2242,65 @@ def split_axis(self, axis, sep='_', names=None, regex=None):
...
+        join : {'outer', 'inner', 'left', 'right'}, optional
+            Defaults to 'outer'.
+        axes : AxisReference or sequence of them, optional
+            Axes to align. Need to be valid in both arrays. Defaults to None (all common axes).
+
+        Returns
+        -------
+        (left, right) : (AxisCollection, AxisCollection)
+            Aligned collections
+
```
Comment (alixdamman) [inline larray/core/axis.py]: See Also
--------
LArray.align

**Comment by alixdamman** [comment]
Couldn't we consider the fact that it doesn't work for anonymous axes as a feature and not a bug? 
I think we should keep it like this for moment and update align if users ask (need).

File: doc/source/changes/version_0_25.rst.inc (hunk 5 above)
Comment (gdementen) [inline doc/source/changes/version_0_25.rst.inc]: there is only one method used at a time

File: larray/core/array.py (hunk 8 above)
Comment (gdementen) [inline larray/core/array.py]: What do you mean by a warning? This is already documented...

Comment (alixdamman) [inline larray/core/array.py]: where? I missed it.

Comment (alixdamman) [inline larray/core/array.py]: where? I missed it.

Comment (alixdamman) [inline larray/core/array.py]: Sorry, for the warning, I was talking about the anonymous axes.

File: larray/core/array.py (hunk 10 above)
Comment (gdementen) [inline larray/core/array.py]: see #335

**Comment by alixdamman** [comment]
Could you add an entry for align in api.rst? Thanks

File: larray/core/axis.py (hunk 13 above)
Comment (alixdamman) [inline larray/core/axis.py]: This docstring should be only read by developers or also users? If developers only, no need for section "See Also".

**Comment by alixdamman** [comment]
LGTM

**Pull request merged**